ai-integrations/langgraph.ipynb

```diff
@@ -0,0 +1,598 @@
+    "from langchain_openai import ChatOpenAI\n",
+    "\n",
+    "# Initialize the LLM\n",
+    "llm = ChatOpenAI()\n",
+    "\n",
+    "# Create a chat prompt template for the agent, which includes a system prompt and a placeholder for `messages`\n",
+    "prompt = ChatPromptTemplate.from_messages(\n",
+    "    [\n",
+    "        (\n",
+    "            \"You are a helpful AI chatbot.\"\n",
+    "            \" You are provided with tools to answer questions about movies.\"\n",
+    "            \" Think step-by-step and use these tools to get the information required to answer the user query.\"\n",
+    "            \" Do not re-run tools unless absolutely necessary.\"\n",
+    "            \" If you are not able to get enough information using the tools, reply with I DON'T KNOW.\"\n",
+    "            \" You have access to the following tools: {tool_names}.\"\n",
+    "        ),\n",
+    "        MessagesPlaceholder(variable_name=\"messages\"),\n",
+    "    ]\n",
+    ")\n",
+    "\n",
+    "# Create the list of tools\n",
+    "tools = [\n",
+    "    vector_search,\n",
+    "    full_text_search\n",
+    "]\n",
+    "\n",
+    "# Partial the prompt template with the tool names\n",
+    "prompt = prompt.partial(tool_names=\", \".join([tool.name for tool in tools]))\n",
+    "\n",
+    "# Bind the `tools` to the `llm` and chain the `prompt` by using the `|` operator\n",
+    "bind_tools = llm.bind_tools(tools)\n",
+    "llm_with_tools = prompt | bind_tools"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Test that the LLM is making the right tool calls\n",
+    "llm_with_tools.invoke([\"What are some movies that take place in the ocean?\"]).tool_calls"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Test that the LLM is making the right tool calls\n",
+    "llm_with_tools.invoke([\"What's the plot of Titanic?\"]).tool_calls"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Build the graph"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from typing import Annotated\n",
+    "from typing_extensions import TypedDict\n",
+    "from langgraph.graph import StateGraph, START\n",
+    "from langgraph.graph.message import add_messages\n",
+    "\n",
+    "# Define the graph state\n",
+    "class GraphState(TypedDict):\n",
+    "    messages: Annotated[list, add_messages]\n",
+    "\n",
+    "# Instantiate the graph\n",
+    "graph = StateGraph(GraphState)\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from typing import Dict, List\n",
+    "\n",
+    "# Define the agent node\n",
+    "def agent(state: GraphState) -> Dict[str, List]:\n",
+    "    \"\"\"\n",
```

Comment:
Similar nit re: the function comment here - this one feels completely unnecessary with the line above explaining what the function is doing, and the arg and return being pretty self-explanatory.